File: extensions/shared/src/main/java/app/morphe/extension/youtube/patches/utils/FreezeLayoutUpdatesPatch.java
package app.morphe.extension.youtube.patches.utils;

import app.morphe.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public final class FreezeLayoutUpdatesPatch {
    /*
     * Ported from kitadai31/revanced-patches-android6-7 Freeze Layout Updates,
     * adapted for Morphe package names and 19.16.39 legacy compatibility.
     */
    private static final boolean ENABLED = Settings.FREEZE_LAYOUT_UPDATES.get();
    private static final boolean DISABLE_LAYOUT_UPDATES = Settings.DISABLE_LAYOUT_UPDATES.get();

    public static String getHotConfigGroup(String original) {
        if (!ENABLED) {
            return original;
        }

        String savedValue = Settings.FROZEN_HOT_CONFIG_GROUP.get();
        if (DISABLE_LAYOUT_UPDATES || savedValue.isEmpty()) {
            return null;
        }
        return savedValue;
    }

    public static String getHotHashData(String original) {
        if (!ENABLED) {
            return original;
        }

        if (DISABLE_LAYOUT_UPDATES) {
            return "";
        }
        return Settings.FROZEN_HOT_HASH_DATA.get();
    }

    public static String getColdConfigGroup(String original) {
        if (!ENABLED) {
            return original;
        }

        String savedValue = Settings.FROZEN_COLD_CONFIG_GROUP.get();
        if (DISABLE_LAYOUT_UPDATES || savedValue.isEmpty()) {
            return null;
        }
        return savedValue;
    }

    public static String getColdHashData(String original) {
        if (!ENABLED) {
            return original;
        }

        if (DISABLE_LAYOUT_UPDATES) {
            return "";
        }
        return Settings.FROZEN_COLD_HASH_DATA.get();
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/general/updates/FreezeLayoutUpdatesPatch.kt
package app.morphe.patches.youtube.general.updates

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.youtube.utils.compatibility.Constants.COMPATIBLE_PACKAGE
import app.morphe.patches.youtube.utils.extension.Constants.UTILS_PATH
import app.morphe.patches.youtube.utils.patch.PatchList.FREEZE_LAYOUT_UPDATES
import app.morphe.patches.youtube.utils.settings.ResourceUtils.addPreference
import app.morphe.patches.youtube.utils.settings.settingsPatch
import app.morphe.util.fingerprint.matchOrThrow
import app.morphe.util.indexOfFirstInstructionOrThrow
import com.android.tools.smali.dexlib2.Opcode

private const val EXTENSION_CLASS_DESCRIPTOR = "$UTILS_PATH/FreezeLayoutUpdatesPatch;"

@Suppress("unused")
val freezeLayoutUpdatesPatch = bytecodePatch(
    FREEZE_LAYOUT_UPDATES.title,
    FREEZE_LAYOUT_UPDATES.summary,
) {
    compatibleWith(COMPATIBLE_PACKAGE)

    dependsOn(settingsPatch)

    execute {
        hotConfigPreferenceFingerprint.matchOrThrow().let { match ->
            match.method.apply {
                val hotConfigGroupResultIndex = match.stringMatches!!.first().index - 1
                addInstructions(
                    hotConfigGroupResultIndex + 1,
                    """
                        invoke-static {v5}, $EXTENSION_CLASS_DESCRIPTOR->getHotConfigGroup(Ljava/lang/String;)Ljava/lang/String;
                        move-result-object v5
                    """
                )

                val hotStoredTimestampResultIndex =
                    indexOfFirstInstructionOrThrow(hotConfigGroupResultIndex, Opcode.MOVE_RESULT_WIDE)

                val hotHashDataResultIndex =
                    indexOfFirstInstructionOrThrow(hotStoredTimestampResultIndex, Opcode.INVOKE_INTERFACE) + 1
                addInstructions(
                    hotHashDataResultIndex + 1,
                    """
                        invoke-static {v2}, $EXTENSION_CLASS_DESCRIPTOR->getHotHashData(Ljava/lang/String;)Ljava/lang/String;
                        move-result-object v2
                    """
                )
            }
        }

        coldConfigPreferenceFingerprint.matchOrThrow().let { match ->
            match.method.apply {
                val coldHashDataResultIndex = match.stringMatches!![2].index + 3
                addInstructions(
                    coldHashDataResultIndex + 1,
                    """
                        invoke-static {v0}, $EXTENSION_CLASS_DESCRIPTOR->getColdHashData(Ljava/lang/String;)Ljava/lang/String;
                        move-result-object v0
                    """
                )

                val coldConfigGroupResultIndex = match.stringMatches!![0].index + 3
                addInstructions(
                    coldConfigGroupResultIndex + 1,
                    """
                        invoke-static {v1}, $EXTENSION_CLASS_DESCRIPTOR->getColdConfigGroup(Ljava/lang/String;)Ljava/lang/String;
                        move-result-object v1
                    """
                )
            }
        }

        addPreference(
            arrayOf(
                "PREFERENCE_SCREEN: GENERAL",
                "PREFERENCE_CATEGORY: GENERAL_EXPERIMENTAL_FLAGS",
                "SETTINGS: FREEZE_LAYOUT_UPDATES"
            ),
            FREEZE_LAYOUT_UPDATES
        )
    }
}
